### ISSUE TITLE
Change working directory on exit?

### ISSUE THREAD

**Issue by Oneechan69** [mikavilpas/yazi.nvim#617]
I want my cwd in nvim to change when I press q in Yazi.nvim, I only quickly looked here but is this possible?

**Comment by mikavilpas** [comment]
Hi, this is not currently supported. As a workaround you can maybe use the `keymaps.change_working_directory` key before you quit yazi.

I could see if I can implement this in the future, so let's keep this open until then.

**Comment by mikavilpas** [comment]
I remembered that this was also discussed in https://github.com/mikavilpas/yazi.nvim/issues/83

It seems to have a custom implementation that you could try using in your own config, if you want to have something working for yourself until I implement this.

**Comment by sxyazi** [comment]
If I understand correctly, you're looking for `yazi --cwd-file`. 

No matter if the user exits Yazi through custom keys or default keys, Yazi will write the last `CWD` to the file specified by `--cwd-file`. 

So, you can pass it when starting Yazi, just like how the [shell wrappers](https://yazi-rs.github.io/docs/quick-start#shell-wrapper) do.

**Comment by mikavilpas** [comment]
Sorry, I forgot to respond earlier and this was autoclosed. The problem before was actually not figuring out what the last directory was, but whether yazi was closed by the user or by yazi.nvim.

Since then, DDS events can be and are used to open files, and with the new addition of https://github.com/sxyazi/yazi/pull/3005 I think I can do a lot of customization if necessary.

Let's give this another shot.

**Comment by mikavilpas** [comment]
I was not able to receive `key-quit` events with `ya sub cd,key-quit`. Maybe a custom plugin is needed to be able to listen to these events right now. Is this correct, @sxyazi ?

I wrote a basic implementation for changing the cwd into the examples - hopefully it can be used to implement this in the future.

**Comment by sxyazi** [comment]
Yes, `key-quit` is a local event rather than a remote one – it needs to execute synchronously and be able to block the event trigger, so it needed to be subscribed within a plugin:

```lua
--- @sync entry
-- ~/.config/yazi/plugins/my-plugin.yazi/main.lua
return {
  entry = function()
    ps.sub("key-quit", function()
      -- ...
    end)
  end
}
```

You can then load and register the event handler with:

```sh
ya emit plugin my-plugin
```

**Comment by mikavilpas** [comment]
Awesome, thanks! I was actually not aware that plugins can be loaded dynamically like this. Maybe I can provide a small plugin like that in yazi.nvim, and it could be activated like that if the user wants to have that feature.

This looks like a good direction to take 👍🏻

**Comment by mikavilpas** [comment]
Ok, this is now possible in the open PR here https://github.com/mikavilpas/yazi.nvim/pull/1087

It requires adding a new yazi plugin `key-event.yazi`, which is bundled in this repository (but still needs to be installed to yazi). It seems to work though!

I'll probably merge it tomorrow or the day after that. If you have time to try it out, let me know what you think @Oneechan69

**Comment by sxyazi** [comment]
I might be wrong, but I looked at the plugin code, and it seems that this is already supported via `yazi --cwd-file` (https://github.com/mikavilpas/yazi.nvim/issues/617#issuecomment-2574707761). Have you encountered any issues with `yazi --cwd-file`?

**Comment by mikavilpas** [comment]
No issues, it might be quite a bit simpler though. The reason I implemented it with a plugin was that I wanted to be sure the `q` key has been pressed, to make a distinction between the user quitting, quitting via a yazi.nvim keymap, and opening a file.

But I realized I could detect a similar situation with `--cwd-file` when no files were opened. I should see if it's possible to use that instead. I have a bit of trouble remembering how the keymaps work, there might be some limitations there

**Comment by mikavilpas** [comment]
It worked, and was much simpler in the end. Good suggestion 👍🏻
